### PR TITLE
Warm reboot: store FDB entries and warm start

### PR DESCRIPTION
**Pull request by qiluo-msft**
Inspired by https://github.com/Azure/sonic-swss/pull/558
The motivation of this PR:

1. orchagent should not directly read ASIC DB
2. orchagent should not assume that syncd recording FDB entries at all. It should store FDB entries based on observation on FDB notifications, and use them during warm start
3. We believe orchagent holds the ground truth of FDB entries during warm start
4. Orchagent will create_fdb_entry for dynamic ones, so sync don’t need special treatment on them during comparison logic
5. Orchagent will receive fdb notifications, mapping bridge_port_id to port name and store it in StateDB for warm restore
6. Orchagent will stop FDB learning/moving/aging before warm stop. (TBD in another PR)

Tested:
1. vs test: store FDB notification in StateDB
2. vs test: warm start with FDB in StateDB

### CODE REVIEW

**Comment by qiluo-msft** [comment]
Retest this please

**Comment by qiluo-msft** [comment]
Retest this please

**Comment by lguohan** [comment]
where do you use refreshFdbEntries?

File: orchagent/orchdaemon.cpp [hunk 1]
```diff
@@ -66,7 +66,11 @@ bool OrchDaemon::init()
 
     gCrmOrch = new CrmOrch(m_configDb, CFG_CRM_TABLE_NAME);
     gPortsOrch = new PortsOrch(m_applDb, ports_tables);
-    gFdbOrch = new FdbOrch(m_applDb, APP_FDB_TABLE_NAME, gPortsOrch);
+
+    TableConnector applDbFdb(m_applDb, APP_FDB_TABLE_NAME);
+    TableConnector stateDbFdb(m_stateDb, APP_FDB_TABLE_NAME);
```
Comment (jipanyang) [inline orchagent/orchdaemon.cpp]: APP_FDB_TABLE_NAME for stateDB?

File: orchagent/fdborch.cpp [hunk 2]
```diff
@@ -36,119 +38,145 @@ FdbOrch::FdbOrch(DBConnector *db, string tableName, PortsOrch *port) :
...
+        this->update(SAI_FDB_EVENT_LEARNED, &fdb_entry, bridge_port_id);
+        SWSS_LOG_INFO("FDB from StateDB %s", key.c_str());
+    }
 
-            if (ret.second)
+    // TODO: do we need to sai_fdb_api->create_fdb_entry() during warm start?
+}
+
+void FdbOrch::update(sai_fdb_event_t type, const sai_fdb_entry_t* fdb_entry, sai_object_id_t bridge_port_id)
+{
```
Comment (jipanyang) [inline orchagent/fdborch.cpp]: Did you make any function change in this function? Any issue with current code?

File: orchagent/orchdaemon.cpp (hunk 1 above)
Comment (qiluo-msft) [reply orchagent/orchdaemon.cpp]: APP_FDB_TABLE_NAME is a table name. FDB_TABLE_NAME may be a better name. I admitted the name is confusing and may refine it in future.

---
In reply to: [217208443](https://github.com/Azure/sonic-swss/pull/615#discussion_r217208443) [](ancestors = 217208443)

File: orchagent/fdborch.cpp (hunk 2 above)
Comment (qiluo-msft) [reply orchagent/fdborch.cpp]: It's just refactoring. No issue. I notice code duplication with FdbOrch::addFdbEntry(). So I make them very similar.


---
In reply to: [217211877](https://github.com/Azure/sonic-swss/pull/615#discussion_r217211877) [](ancestors = 217211877)

**Comment by qiluo-msft** [comment]
Retest this please

**Comment by qiluo-msft** [comment]
To answer @jipanyang's question
> How are the FDB entries retstored? It looks FDB entries are saved in stateDB.
> Could you separate code restructuring with the function implementation in general?
> There are quite some changes not related to FDB restore.

The function refreshFdbEntries() will restore FDB from StateDB. The real actions are actually deserializing and update observing orchs.
Other code changes are just refactoring.

File: orchagent/fdborch.cpp [hunk 3]
```diff
@@ -36,119 +38,145 @@ FdbOrch::FdbOrch(DBConnector *db, string tableName, PortsOrch *port) :
     Orch::addExecutor(fdbNotifier);
 }
 
-void FdbOrch::update(sai_fdb_event_t type, const sai_fdb_entry_t* entry, sai_object_id_t bridge_port_id)
+void FdbOrch::refreshFdbEntries()
```
[the remark this answers is not in the document]
Comment (lguohan) [reply orchagent/fdborch.cpp]: we plan to this reconciliation by disabling hw learn during warm-reboot on the asic.

**Comment by jipanyang** [comment]
@qiluo-msft I didn't find the place  where refreshFdbEntries() is called.   Mixing code refactoring with new feature implementation makes it pretty hard to do code review.

@lguohan "we plan to this reconciliation by disabling hw learn during warm-reboot on the asic",  in this case,  unplanned warm restart won't work.  swss won't be able to signal asic for unplanned restart.

**Comment by lguohan** [comment]
@jipanyang , unplanned restart is not a goal here. It is very difficult to recover the system start from a crashed scenario.

**Comment by lguohan** [comment]
@qiluo-msft , please indicate us to do the code review when you use the refreshfdbentries, both jipan and me asked the same question. not much to review for now.

**Comment by qiluo-msft** [comment]
In latest iteration, all warm start logic is moved into bake(), which called by orchdaemon.

---
In reply to: [420544357](https://github.com/Azure/sonic-swss/pull/615#issuecomment-420544357) [](ancestors = 420544357)

**Comment by qiluo-msft** [comment]
In latest iteration, all warm start logic is moved into bake(), which called by orchdaemon. Refactoring removed.

---
In reply to: [421503216](https://github.com/Azure/sonic-swss/pull/615#issuecomment-421503216) [](ancestors = 421503216)

**Comment by qiluo-msft** [comment]
Please help review and meanwhile I will fix the unit test.

File: orchagent/fdborch.cpp [hunk 4]
```diff
@@ -36,6 +37,73 @@ FdbOrch::FdbOrch(DBConnector *db, string tableName, PortsOrch *port) :
...
+        {
+            SWSS_LOG_INFO("FdbOrch notification: mac %s is duplicate", entry.mac.to_string().c_str());
+            return false;
+        }
+
+        // Write to StateDb
+        std::vector<FieldValueTuple> fvs;
+        fvs.push_back(FieldValueTuple("port", portName));
+        fvs.push_back(FieldValueTuple("type", "dynamic"));
+        m_fdbStateTable.set(key, fvs);
```
Comment (jipanyang) [inline orchagent/fdborch.cpp]: If FDB entries are saved to stateDB. During system warm reboot,  will stateDB be restored?

Comment (qiluo-msft) [reply orchagent/fdborch.cpp]: Do you mean will stateDB in redis persistent during warm reboot? Yes, it should be.

Or do you mean if fdborch will restore from stateDB? Yes, restored in bake().

---
In reply to: [218964895](https://github.com/Azure/sonic-swss/pull/615#discussion_r218964895) [](ancestors = 218964895)

File: orchagent/orchdaemon.cpp [hunk 5]
```diff
@@ -454,7 +458,7 @@ bool OrchDaemon::warmRestoreValidation()
         }
     }
     WarmStart::setWarmStartState("orchagent", WarmStart::RESTORED);
-    return true;
+    return ts.empty();
```
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply orchagent/orchdaemon.cpp]: If ts empty, warm restore is successful.

---
In reply to: [218952757](https://github.com/Azure/sonic-swss/pull/615#discussion_r218952757) [](ancestors = 218952757)

File: orchagent/fdborch.cpp (hunk 4 above)
Comment (lguohan) [reply orchagent/fdborch.cpp]: no stateDB will be cleared during state warm reboot. it is a problem.

Comment (lguohan) [reply orchagent/fdborch.cpp]: I will save the fdb entries in state db in the warm-reboot script and recover it in the going up path. @jipanyang , thanks for the catch.

Comment (jipanyang) [reply orchagent/fdborch.cpp]: Would it be easier if the data is put in a separate table in appDB so no special handling is needed for stateDB?

Comment (jipanyang) [reply orchagent/fdborch.cpp]: Would it be easier if the data is put in a separate table in appDB so no special handling is needed for stateDB?

File: orchagent/orchdaemon.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jipanyang) [reply orchagent/orchdaemon.cpp]: Got it, thanks

File: orchagent/orchdaemon.cpp (hunk 1 above)
Comment (qiluo-msft) [reply orchagent/orchdaemon.cpp]: Resolved in this PR.

---
In reply to: [217835174](https://github.com/Azure/sonic-swss/pull/615#discussion_r217835174) [](ancestors = 217835174,217208443)

**Comment by qiluo-msft** [comment]
Retest this please

File: tests/test_fdb.py [hunk 6]
```diff
@@ -122,10 +86,114 @@ def test_FDBAddedAfterMemberCreated(dvs):
...
+        # Get mapping between interface name and its bridge port_id
+        # Note: they are changed
+        iface_2_bridge_port_id = get_map_iface_bridge_port_id(dvs.adb, dvs)
+
+        # check that the FDB entries were inserted into ASIC DB
+        ok, extra = dvs.is_fdb_entry_exists(dvs.adb, "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY",
+                        [],
+                        [("SAI_FDB_ENTRY_ATTR_TYPE", "SAI_FDB_ENTRY_TYPE_DYNAMIC"),
+                         ("SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID", iface_2_bridge_port_id["Ethernet64"]),
+                        ]
```
Comment (jipanyang) [inline tests/test_fdb.py]: Was the entry removed during orchagent restart?  If not,  probably other check like CRM counter value might be better.

File: tests/test_fdb.py [hunk 7]
```diff
@@ -122,10 +86,114 @@ def test_FDBAddedAfterMemberCreated(dvs):
...
+                        [],
+                        [("SAI_FDB_ENTRY_ATTR_TYPE", "SAI_FDB_ENTRY_TYPE_DYNAMIC"),
+                         ("SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID", iface_2_bridge_port_id["Ethernet68"]),
+                        ]
+        )
+        assert ok, str(extra)
+
+    finally:
+        # disable warm restart
+        # TODO: use cfg command to config it
```
Comment (jipanyang) [inline tests/test_fdb.py]: "config warm_restart enable swss" is available.

File: tests/test_fdb.py [hunk 8]
```diff
@@ -122,10 +86,114 @@ def test_FDBAddedAfterMemberCreated(dvs):
...
+        dvs.cdb,
+        swsscommon.CFG_WARM_RESTART_TABLE_NAME, "swss",
+        [
+            ("enable", "true"),
+        ]
+    )
+
+    try:
+        # restart orchagent
+        dvs.runcmd(['sh', '-c', 'supervisorctl restart orchagent'])
```
Comment (jipanyang) [inline tests/test_fdb.py]: Would it be better to do swss start/stop?   If it is for warm restart testing,  test_warm_reboot.py might be a good place for all related test cases.    
```
# start processes in SWSS
def start_swss(dvs):
    dvs.runcmd(['sh', '-c', 'supervisorctl start orchagent; supervisorctl start portsyncd; supervisorctl start intfsyncd; \
        supervisorctl start neighsyncd; supervisorctl start intfmgrd; supervisorctl start vlanmgrd; \
        supervisorctl start buffermgrd; supervisorctl start arp_update'])

# stop processes in SWSS
def stop_swss(dvs):
    dvs.runcmd(['sh', '-c', 'supervisorctl stop orchagent; supervisorctl stop portsyncd; supervisorctl stop intfsyncd; \
        supervisorctl stop neighsyncd;  supervisorctl stop intfmgrd; supervisorctl stop vlanmgrd; \
        supervisorctl stop buffermgrd; supervisorctl stop arp_update'])

```

File: orchagent/fdborch.cpp [hunk 9]
```diff
@@ -36,6 +37,73 @@ FdbOrch::FdbOrch(DBConnector *db, string tableName, PortsOrch *port) :
...
+bool FdbOrch::storeFdbEntryState(const FdbUpdate& update)
+{
+    const FdbEntry& entry = update.entry;
+    const Port& port = update.port;
+    sai_vlan_id_t vlan_id = port.m_port_vlan_id;
+    const MacAddress& mac = entry.mac;
+    string portName = port.m_alias;
+
+    // ref: https://github.com/Azure/sonic-swss/blob/master/doc/swss-schema.md#fdb_table
+    string key = "Vlan" + to_string(vlan_id) + ":" + mac.to_string();
```
Comment (lguohan) [inline orchagent/fdborch.cpp]: >: [](start = 48, length = 1)

should this be '|'

File: orchagent/fdborch.cpp [hunk 10]
```diff
@@ -36,6 +37,73 @@ FdbOrch::FdbOrch(DBConnector *db, string tableName, PortsOrch *port) :
...
+    auto consumer = dynamic_cast<Consumer *>(getExecutor(APP_FDB_TABLE_NAME));
+    if (consumer == NULL)
+    {
+        SWSS_LOG_ERROR("No consumer %s in Orch", APP_FDB_TABLE_NAME);
+        return false;
+    }
+
+    size_t refilled = consumer->refillToSync(&m_fdbStateTable);
+    SWSS_LOG_NOTICE("Add warm input FDB State: %s, %zd", APP_FDB_TABLE_NAME, refilled);
+    return true;
```
Comment (lguohan) [inline orchagent/fdborch.cpp]: does this call fdb notification? here the consumer is the app db? how can the fdb entries in state db got translated into fdb notificastions?

File: tests/test_fdb.py (hunk 6 above)
Comment (qiluo-msft) [reply tests/test_fdb.py]: If you are asking entry in ASIC DB, no. orchagent warm start will create dynamic fdb, and syncd apply view will match it with old view, and keep it.
I will add CRM test.

---
In reply to: [219020615](https://github.com/Azure/sonic-swss/pull/615#discussion_r219020615) [](ancestors = 219020615)

File: tests/test_fdb.py (hunk 7 above)
Comment (qiluo-msft) [reply tests/test_fdb.py]: I am following tests in test_warm_reboot.py. I will update later if those tests updated.

---
In reply to: [219020740](https://github.com/Azure/sonic-swss/pull/615#discussion_r219020740) [](ancestors = 219020740)

File: orchagent/fdborch.cpp (hunk 9 above)
Comment (qiluo-msft) [reply orchagent/fdborch.cpp]: Currently there is a trick. All entries stored in StateDB FDB_TABLE will be fed into ApplDB during warm starting. So I keep them the same pattern.

---
In reply to: [219357862](https://github.com/Azure/sonic-swss/pull/615#discussion_r219357862) [](ancestors = 219357862)

File: orchagent/fdborch.cpp (hunk 10 above)
Comment (qiluo-msft) [reply orchagent/fdborch.cpp]: 1. Are you asking syncd receive this fdb entry and send orchagent notification? No.
2. Yes, the consumer is the app db
3. Never translated.

---
In reply to: [219421390](https://github.com/Azure/sonic-swss/pull/615#discussion_r219421390) [](ancestors = 219421390)

File: orchagent/fdborch.cpp (hunk 9 above)
Comment (jipanyang) [reply orchagent/fdborch.cpp]: I still feel it is more convenient to save the data in appDB using a new table like ASIC_FDB_TABLE, so to eliminate all the exception handling (stateDB save/restore during system warm reboot, and the special ":" separator in stateDB).

**Comment by qiluo-msft** [comment]
Retest this please

**Pull request merged**